src/Microsoft.Diagnostics.Monitoring.RestServer/Controllers/DiagController.cs

```diff
@@ -79,20 +82,67 @@ public Task<ActionResult> GetGcDump(int? pid)
             [FromQuery]TraceProfile profile = DefaultTraceProfiles,
-            [FromQuery][Range(-1, int.MaxValue)]int durationSeconds = 30,
+            [FromQuery][Range(-1, int.MaxValue)] int durationSeconds = 30,
             [FromQuery][Range(1, int.MaxValue)] int metricsIntervalSeconds = 1)
         {
             TimeSpan duration = ConvertSecondsToTimeSpan(durationSeconds);
-            return InvokeService(async () =>
+
+            var configurations = new List<MonitoringSourceConfiguration>();
+            if (profile.HasFlag(TraceProfile.Cpu))
             {
-                int pidValue = _diagnosticServices.ResolveProcess(pid);
-                IStreamWithCleanup result = await _diagnosticServices.StartTrace(pidValue, profile, duration, metricsIntervalSeconds, this.HttpContext.RequestAborted);
-                return new StreamWithCleanupResult(result, "application/octet-stream", FormattableString.Invariant($"{Guid.NewGuid()}.nettrace"));
-            });
+                configurations.Add(new CpuProfileConfiguration());
+            }
+            if (profile.HasFlag(TraceProfile.Http))
+            {
+                configurations.Add(new HttpRequestSourceConfiguration());
+            }
+            if (profile.HasFlag(TraceProfile.Logs))
+            {
+                configurations.Add(new LoggingSourceConfiguration());
+            }
+            if (profile.HasFlag(TraceProfile.Metrics))
+            {
+                configurations.Add(new MetricSourceConfiguration(metricsIntervalSeconds));
+            }
+
+            var aggregateConfiguration = new AggregateSourceConfiguration(configurations.ToArray());
+
+            return StartTrace(pid, aggregateConfiguration, duration);
+        }
+
+        [HttpPost("trace/{pid?}")]
+        public Task<ActionResult> TracePostedConfiguration(
+            int? pid,
+            [FromBody][Required] EventPipeConfigurationModel configuration,
+            [FromQuery][Range(-1, int.MaxValue)] int durationSeconds = 30)
+        {
+            TimeSpan duration = ConvertSecondsToTimeSpan(durationSeconds);
+
+            var providers = new List<EventPipeProvider>();
+
+            foreach (EventPipeProviderModel providerModel in configuration.Providers)
+            {
+                long keywords = long.Parse(providerModel.Keywords, NumberStyles.HexNumber, CultureInfo.InvariantCulture);
+                EventLevel eventLevel = MapEventLevel(providerModel.EventLevel);
+
+                providers.Add(new EventPipeProvider(
+                    providerModel.Name,
+                    eventLevel,
+                    keywords,
+                    providerModel.Arguments
+                    ));
+            }
+
+            var traceConfiguration = new EventPipeProviderSourceConfiguration(
+                providers: providers.ToArray(),
+                requestRundown: configuration.RequestRundown,
+                bufferSizeInMB: configuration.BufferSizeInMB);
```

Comment:
Does the underlying EventPipe have an upper bound on BufferSizeInMB? If not, we should consider one.

Reply:
I've been looking through the docs and code to see if there's an upper limit and it doesn't seem to have an artificial one beyond what the type can allow (the runtime uses uint/uint32_t whereas the diagnostics repo uses int). The runtime does require that that the buffer is greater than 0, so we can enforce that lower bound ahead of time.